pkg/service/tools/tools.go

```diff
@@ -294,3 +319,143 @@ func (t *Tools) IgnoreTests(_ context.Context, _ string, _ []string) error {
+	if err != nil {
+		t.logger.Error("Error requesting device code", zap.Error(err))
+		return false
+	}
+
+	promptUser(deviceCodeResp)
+
+	tokenResp, err := pollForAccessToken(ctx, t.logger, deviceCodeResp.DeviceCode, deviceCodeResp.Interval)
+	if err != nil {
+		if ctx.Err() == context.Canceled {
+			return false
+		}
+		utils.LogError(t.logger, err, "failed to poll for access token")
+		return false
+	}
+
+	userEmail, isValid, _, authErr, err := utils.CheckAuth(ctx, utils.APIServerURL, tokenResp.AccessToken, true, t.logger)
+	if err != nil {
+		if ctx.Err() == context.Canceled {
+			return false
+		}
+		t.logger.Error("Error checking auth", zap.Error(err))
+		return false
+	}
+
+	if !isValid {
+		t.logger.Error("Invalid token", zap.Any("error", authErr))
+		return false
+	}
+
+	err = t.configdb.WriteAccessToken(ctx, tokenResp.AccessToken)
+	if err != nil {
+		if ctx.Err() == context.Canceled {
+			return false
+		}
+		t.logger.Error("Error writing access token", zap.Error(err))
+		return false
+	}
+	fmt.Println("Successfully logged in to Keploy using GitHub as " + userEmail)
+	return true
+}
+
+func requestDeviceCode(logger *zap.Logger) (*DeviceCodeResponse, error) {
+	data := url.Values{}
+	data.Set("client_id", clientID)
+	data.Set("scope", "read:user")
+
+	resp, err := http.PostForm(deviceCodeURL, data)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err := resp.Body.Close()
+		if err != nil {
+			utils.LogError(logger, err, "failed to close response body")
+		}
+	}()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse the URL-encoded response
+	parsed, err := url.ParseQuery(string(body))
+	if err != nil {
+		return nil, err
+	}
+
+	// Populate the DeviceCodeResponse struct
+	deviceCodeResp := &DeviceCodeResponse{
+		DeviceCode:      parsed.Get("device_code"),
+		UserCode:        parsed.Get("user_code"),
+		VerificationURI: parsed.Get("verification_uri"),
+		Interval:        5, // Default value; you can parse this from the response if needed
+	}
+
+	return deviceCodeResp, nil
+}
+
+func promptUser(deviceCodeResp *DeviceCodeResponse) {
+	fmt.Printf("Please go to %s and enter the code: %s\n", deviceCodeResp.VerificationURI, deviceCodeResp.UserCode)
+}
+
+func pollForAccessToken(ctx context.Context, logger *zap.Logger, deviceCode string, interval int) (*AccessTokenResponse, error) {
+	data := url.Values{}
+	data.Set("client_id", clientID)
+	data.Set("device_code", deviceCode)
+	data.Set("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
+
+	fmt.Println("waiting for approval...")
+
+	for {
+		resp, err := http.PostForm(tokenURL, data)
+		if err != nil {
+			return nil, err
+		}
+		defer func() {
+			err := resp.Body.Close()
+			if err != nil {
+				utils.LogError(logger, err, "failed to close response body")
+			}
+		}()
+
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		if resp.StatusCode == http.StatusOK {
```

Comment:
should invert this

Reply:
are saying about early return if there is non success code..?

Reply:
done